providers/digitalocean/digitalocean.go

```diff
@@ -0,0 +1,189 @@
+}
+
+type TokenSource struct {
+	AccessToken string
+}
+
+func (t *TokenSource) Token() (*oauth2.Token, error) {
+	token := &oauth2.Token{
+		AccessToken: t.AccessToken,
+	}
+	return token, nil
+}
+
+func (p *DigitalOceanProvider) Init(rootDomainName string) error {
+	var pat string
+	if pat = os.Getenv("DO_PAT"); len(pat) == 0 {
+		return fmt.Errorf("DO_PAT is not set")
+	}
+
+	tokenSource := &TokenSource{
+		AccessToken: pat,
+	}
+
+	oauthClient := oauth2.NewClient(oauth2.NoContext, tokenSource)
+	p.client = api.NewClient(oauthClient)
+
+	// DO's API is rate limited at 5000/hour.
+	doqps := (float64)(5000.0 / 3600.0)
+	p.limiter = ratelimit.NewBucketWithRate(doqps, 1)
+
+	p.rootDomainName = utils.UnFqdn(rootDomainName)
+
+	// Retrieve email address associated with this PAT.
+	p.limiter.Wait(1)
+	acct, _, err := p.client.Account.Get()
+	if err != nil {
+		return err
+	}
+
+	// Now confirm that domain is accessible under this PAT.
+	p.limiter.Wait(1)
+	domains, _, err := p.client.Domains.Get(p.rootDomainName)
+	if err != nil {
+		return err
+	}
+	// DO's TTLs are domain-wide.
+	p.TTL = domains.TTL
+
+	logrus.Infof("Configured %s for email %s and domain %s", p.GetName(), acct.Email, domains.Name)
+
+	return nil
+}
+
+func (p *DigitalOceanProvider) GetName() string {
+	return "DigitalOcean"
+}
+
+func (p *DigitalOceanProvider) HealthCheck() error {
+	p.limiter.Wait(1)
+	_, _, err := p.client.Domains.Get(p.rootDomainName)
+	return err
+}
+
+func (p *DigitalOceanProvider) AddRecord(record utils.DnsRecord) error {
+	logrus.Debugf("AddRecord")
+	for _, r := range record.Records {
+		createRequest := &api.DomainRecordEditRequest{
+			Type: record.Type,
+			Name: record.Fqdn,
+			Data: r,
+		}
+		logrus.Debugf(" request: %v", createRequest)
+		p.limiter.Wait(1)
+		rec, _, err := p.client.Domains.CreateRecord(p.rootDomainName, createRequest)
+		if err != nil {
+			return fmt.Errorf("%s API call has failed: %v", p.GetName(), err)
+		}
+		logrus.Debugf(" rec: %v", rec)
+	}
+	return nil
+}
+
+func (p *DigitalOceanProvider) UpdateRecord(record utils.DnsRecord) error {
+	logrus.Debugf("UpdateRecord")
+	if err := p.RemoveRecord(record); err != nil {
+		return err
+	}
+	return p.AddRecord(record)
+}
+
+func (p *DigitalOceanProvider) RemoveRecord(record utils.DnsRecord) error {
+	logrus.Debugf("RemoveRecord")
+	p.limiter.Wait(1)
+	records, _, err := p.client.Domains.Records(p.rootDomainName, nil)
+	if err != nil {
+		return err
+	}
+	for _, rec := range records {
+		if rec.Name == record.Fqdn && rec.Type == record.Type {
+			p.limiter.Wait(1)
+			_, err := p.client.Domains.DeleteRecord(p.rootDomainName, rec.ID)
+			if err != nil {
+				return fmt.Errorf("%s API call has failed: %v", p.GetName(), err)
+			}
+		}
+	}
+	return err
+}
+
+func (p *DigitalOceanProvider) GetRecords() ([]utils.DnsRecord, error) {
+	dnsRecords := []utils.DnsRecord{}
+	recordMap := map[string]map[string][]string{}
+	opt := &api.ListOptions{}
```

Comment:
This is a great idea, and should alleviate any sort of bottleneck seen by the folks mentioned in the previous comment.